# Web Scrapping from Autoline website

Autoline website which contains the details of used `cars` around the mentioned city is used for scrapping. 

This method of scrapping uses `cURL` for sending HTTP requests to retrieve the data from the website.

## Importing Libraries

In [1]:
import requests
import pandas as pd

## Requesting connection for the first city

In [2]:
headers = {
    'Accept': '*/*',
    'Accept-Language': 'en-GB,en;q=0.9,ta-IN;q=0.8,ta;q=0.7,en-US;q=0.6',
    'Connection': 'keep-alive',
    'If-None-Match': 'W/^\\^e3211357bab550aa0622b8cbd5cf050e^\\^',
    'Origin': 'https://www.autolist.com',
    'Referer': 'https://www.autolist.com/',
    'Sec-Fetch-Dest': 'empty',
    'Sec-Fetch-Mode': 'cors',
    'Sec-Fetch-Site': 'same-site',
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36',
    'sec-ch-ua': '^\\^Not.A/Brand^\\^;v=^\\^8^\\^, ^\\^Chromium^\\^;v=^\\^114^\\^, ^\\^Google',
    'sec-ch-ua-mobile': '?0',
    'sec-ch-ua-platform': '^\\^Windows^\\^',
}

params = (
    ('city', 'California^%^20City'),
    ('latitude', '35.125801'),
    ('limit', '20'),
    ('longitude', '-117.9859038'),
    ('make', 'Ford'),
    ('page', '1'),
    ('radius', '100'),
    ('state', 'CA'),
)

response = requests.get('https://production-proxy.autolist.com/api/cwv/seo/listings', headers=headers, params=params)

The cURL command for is give to send a HTTP request using requests library to the Autoline website.

In [3]:
response

<Response [200]>

This `<Response [200]>` means that a successfull connection is made to the Website for retrieving the data

In [4]:
response.json()

{'total_results_count': 10619,
 'make_model_name': 'Ford',
 'trims': [],
 'ad_info': {'ad_unit': '/19485787/AutoList.com/Search_Results_Page/Ford',
  'make': 'Ford',
  'model': None,
  'page_type': 'srp',
  'cpo': False,
  'body_style': None,
  'year': 2024,
  'age_type': 'new',
  'app': 'web',
  'seller_id': None,
  'content_url': 'https://www.autolist.com/listings#city=California%5E%25%5E20City&latitude=35.125801&limit=20&longitude=-117.9859038&make=Ford&page=1&radius=100&state=CA'},
 'search_results': [{'available_nationwide': False,
   'created_at': 1680711999,
   'recent_price_drop': False,
   'is_hot': True,
   'href_target': '/listings/3FA6P0LU1JR206181',
   'vdp_url': '/ford-fusion+hybrid#vin=3FA6P0LU1JR206181',
   'clickoff_url': 'www.montclairautox.com?utm_source=autolist&utm_medium=referral&utm_campaign=vdp_deep_link',
   'accepts_leads': True,
   'open_in_new_window': False,
   'id': 262679309,
   'primary_photo_url': 'https://static.cargurus.com/images/forsale/2023/04/17/2

In [5]:
data_json=response.json()

In [6]:
type(data_json)

dict

In [7]:
data_json.keys()

dict_keys(['total_results_count', 'make_model_name', 'trims', 'ad_info', 'search_results', 'html'])

The required datas to be scrapped off are
- Model
- Year
- Dealer
- Mileage
- Price
- City

## Scrapping the necessary datas from the website

In [8]:
#data Scrapping

In [9]:
data_json['search_results']

[{'available_nationwide': False,
  'created_at': 1680711999,
  'recent_price_drop': False,
  'is_hot': True,
  'href_target': '/listings/3FA6P0LU1JR206181',
  'vdp_url': '/ford-fusion+hybrid#vin=3FA6P0LU1JR206181',
  'clickoff_url': 'www.montclairautox.com?utm_source=autolist&utm_medium=referral&utm_campaign=vdp_deep_link',
  'accepts_leads': True,
  'open_in_new_window': False,
  'id': 262679309,
  'primary_photo_url': 'https://static.cargurus.com/images/forsale/2023/04/17/23/15/2018_ford_fusion_hybrid-pic-3193442773663903549-1024x768.jpeg',
  'thumbnail_url_large': 'https://images.autolist.com/image/fetch/s--Ke9T4JwJ--/c_fill,f_auto,q_auto,w_320/https://static.cargurus.com/images/forsale/2023/04/17/23/15/2018_ford_fusion_hybrid-pic-3193442773663903549-1024x768.jpeg',
  'lat': 34.0643,
  'lon': -117.69,
  'year': 2018,
  'make': 'Ford',
  'model': 'Fusion Hybrid',
  'model_id': 959,
  'display_color': '',
  'price': '$18,502',
  'price_unformatted': 18502,
  'mileage': '52,618 Miles',

In [10]:
type(data_json['search_results'])

list

In [11]:
len(data_json['search_results'])

20

As each page contains 20 entries, the length is shown as `20`

In [12]:
data_json['search_results'][0]

{'available_nationwide': False,
 'created_at': 1680711999,
 'recent_price_drop': False,
 'is_hot': True,
 'href_target': '/listings/3FA6P0LU1JR206181',
 'vdp_url': '/ford-fusion+hybrid#vin=3FA6P0LU1JR206181',
 'clickoff_url': 'www.montclairautox.com?utm_source=autolist&utm_medium=referral&utm_campaign=vdp_deep_link',
 'accepts_leads': True,
 'open_in_new_window': False,
 'id': 262679309,
 'primary_photo_url': 'https://static.cargurus.com/images/forsale/2023/04/17/23/15/2018_ford_fusion_hybrid-pic-3193442773663903549-1024x768.jpeg',
 'thumbnail_url_large': 'https://images.autolist.com/image/fetch/s--Ke9T4JwJ--/c_fill,f_auto,q_auto,w_320/https://static.cargurus.com/images/forsale/2023/04/17/23/15/2018_ford_fusion_hybrid-pic-3193442773663903549-1024x768.jpeg',
 'lat': 34.0643,
 'lon': -117.69,
 'year': 2018,
 'make': 'Ford',
 'model': 'Fusion Hybrid',
 'model_id': 959,
 'display_color': '',
 'price': '$18,502',
 'price_unformatted': 18502,
 'mileage': '52,618 Miles',
 'mileage_unformatted

All the necessary datas for the first item is accessed and printed

## Accessing the retrieved data

In [13]:
data_json['search_results'][0]['model']

'Fusion Hybrid'

In [14]:
data_json['search_results'][0]['mileage']

'52,618 Miles'

In [15]:
data_json['search_results'][0]['year']

2018

In [16]:
data_json['search_results'][0]['dealer_name']

'Montclair Auto Exchange'

In [17]:
data_json['search_results'][0]['price']

'$18,502'

For accessing the entire items in the first page, `For` loop is used and the scrapped details are stored as `List`

In [18]:
#looping the information
data_items=data_json['search_results']

In [19]:
model = []
year = []
dealer = []
mileage = []
price = []
city =[]

for data in data_items:
    model.append(data['model'])
    year.append(data['year'])
    dealer.append(data['dealer_name'])
    mileage.append(data['mileage'])
    price.append(data['price'])
    city.append(data['city'])

In [20]:
model

['Fusion Hybrid',
 'Fusion Energi',
 'Escape',
 'Escape',
 'Fusion',
 'Escape',
 'Flex',
 'Fusion',
 'Mustang',
 'Mustang',
 'Taurus',
 'Fusion Hybrid',
 'Flex',
 'Escape',
 'Mustang',
 'Flex',
 'F-150',
 'Escape',
 'Escape Plug-In Hybrid',
 'Escape']

The first 20 car model names scrapped from the website are displayed 

This website totally contains the details of used cars in `633` pages. Hence `multiple pages` are to be scrapped by using the same cURL

## Multiple page Web Scrapping

In [21]:
#scrapping multiple webpages
headers = {
    'Accept': '*/*',
    'Accept-Language': 'en-GB,en;q=0.9,ta-IN;q=0.8,ta;q=0.7,en-US;q=0.6',
    'Connection': 'keep-alive',
    'If-None-Match': 'W/^\\^e3211357bab550aa0622b8cbd5cf050e^\\^',
    'Origin': 'https://www.autolist.com',
    'Referer': 'https://www.autolist.com/',
    'Sec-Fetch-Dest': 'empty',
    'Sec-Fetch-Mode': 'cors',
    'Sec-Fetch-Site': 'same-site',
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36',
    'sec-ch-ua': '^\\^Not.A/Brand^\\^;v=^\\^8^\\^, ^\\^Chromium^\\^;v=^\\^114^\\^, ^\\^Google',
    'sec-ch-ua-mobile': '?0',
    'sec-ch-ua-platform': '^\\^Windows^\\^',
}

model = []
year = []
dealer = []
mileage = []
price = []
city=[]

for i in range(1,634):
    
    params = (
        ('city', 'California^%^20City'),
        ('latitude', '35.125801'),
        ('limit', '20'),
        ('longitude', '-117.9859038'),
        ('make', 'Ford'),
        ('page', str(i)),
        ('radius', '100'),
        ('state', 'CA'),
    )
    #variable for obtaining request
    response = requests.get('https://production-proxy.autolist.com/api/cwv/seo/listings', headers=headers, params=params)
    
    #json object
    data_json=response.json()
    
    #getting 20 items for each page
    data_items=data_json['search_results']
    
    for data in data_items:
        model.append(data['model'])
        year.append(data['year'])
        dealer.append(data['dealer_name'])
        mileage.append(data['mileage'])
        price.append(data['price'])
        city.append(data['city'])
    

After scrapping off entire pages, the retrieved datas are stored in a `DataFrame` format named as `ford_pages_df`.

## Converting into DataFrame

In [22]:
ford_pages_df=pd.DataFrame({'Model': model,'Mileage': mileage,'City': city,'Year': year,'Dealer': dealer,'Price': price})

In [23]:
ford_pages_df

Model       Mileage        City  Year  \
0      Fusion Hybrid  52,618 Miles   Montclair  2018   
1      Fusion Energi  68,878 Miles   Montclair  2017   
2             Escape  20,032 Miles       Norco  2021   
3             Escape  60,292 Miles   Montclair  2018   
4             Fusion  54,599 Miles       Azusa  2018   
...              ...           ...         ...   ...   
9995          Bronco           New  Long Beach  2023   
9996  Mustang Mach-E           New    Redlands  2023   
9997           F-150           New    Redlands  2023   
9998  Mustang Mach-E           New      Oxnard  2023   
9999          Escape           New    Glendale  2023   

                                                 Dealer             Price  
0                               Montclair Auto Exchange           $18,502  
1                               Montclair Auto Exchange           $17,413  
2     CarMax Norco - Now Offering Curbside Pickup & ...           $26,998  
3                               Montclair Auto Exchange           $16,324  
4                              California Motor Trade 2           $16,499  
...                                                 ...               ...  
9995                                 Ford Of Long Beach  accepting_offers  
9996                         Ken Grody Ford Of Redlands  accepting_offers  
9997                         Ken Grody Ford Of Redlands           $68,210  
9998                    Envision Ford Lincoln Of Oxnard  accepting_offers  
9999                                  Star Ford Lincoln  accepting_offers  

[10000 rows x 6 columns]

The entire scrapped datas are a count of `10000` cars from `633` pages

Now, second city is choosen for scrpping the car details

Before that, a HTTP request is give for retrieving the data from the website using `requests` library

## Requesting connection for the second city

In [24]:
#second city basic connection establishing
headers = {
    'sec-ch-ua': '^\\^Not.A/Brand^\\^;v=^\\^8^\\^, ^\\^Chromium^\\^;v=^\\^114^\\^, ^\\^Google',
    'Referer': 'https://www.autolist.com/',
    'sec-ch-ua-mobile': '?0',
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36',
    'sec-ch-ua-platform': '^\\^Windows^\\^',
}

params = (
    ('include_total_price_change', 'true'),
    ('include_time_on_market', 'true'),
    ('include_relative_price_difference', 'true'),
    ('latitude', '42.8286091'),
    ('limit', '20'),
    ('longitude', '-98.3298021'),
    ('make', 'Ford'),
    ('page', '1'),
    ('radius', '100'),
    ('zip', '600099'),
)

response = requests.get('https://production-proxy.autolist.com/api/v2/search', headers=headers, params=params)

In [25]:
response

<Response [200]>

This `<Response [200]>` means that a successfull connection is made to the Website for retrieving the data

In [26]:
response.json()

{'total_count': 3621,
 'total_count_formatted': '3,621',
 'hits_count': 20,
 'records': [{'accepts_click_off': False,
   'accepts_leads': True,
   'accepts_leads_and_click_off': False,
   'always_ask_for_zip': False,
   'available_nationwide': False,
   'backup_camera': True,
   'bed': None,
   'bluetooth': False,
   'body_style': 'suv',
   'body_type': 'suv',
   'cabin': None,
   'categories': ['american'],
   'city': 'Sioux Falls',
   'city_mpg': 19,
   'clickoff_url': '',
   'combined_mpg': 22,
   'condition': 'used',
   'cpl_value': 3,
   'created_at': '2023-06-30T02:01:50.807Z',
   'dealer_group_uuid': None,
   'dealer_name': 'Billion Chrysler Jeep Dodge Ram',
   'distance_from_origin': 91,
   'door_count': 4,
   'driveline': 'FWD',
   'eligible_for_financing': False,
   'email_opt_default': True,
   'engine_cylinders': 6,
   'entertainment': False,
   'experience': 'local',
   'finance_dealer_id': None,
   'financing_experience': 'none',
   'fuel_type': 'gasoline',
   'green_engi

In [27]:
data_jason=response.json()

In [28]:
data_jason.keys()

dict_keys(['total_count', 'total_count_formatted', 'hits_count', 'records', 'promoted_aggregations', 'facet_counts'])

In [29]:
len(data_jason['records'])

20

The carlist in the first page is `20`, so it shows the length as 20

The required datas to be scrapped off are
- Model
- Year
- Dealer
- Mileage
- Price
- City

## Scrapping the necessary datas from the website

In [30]:
data_jason['records'][0]

{'accepts_click_off': False,
 'accepts_leads': True,
 'accepts_leads_and_click_off': False,
 'always_ask_for_zip': False,
 'available_nationwide': False,
 'backup_camera': True,
 'bed': None,
 'bluetooth': False,
 'body_style': 'suv',
 'body_type': 'suv',
 'cabin': None,
 'categories': ['american'],
 'city': 'Sioux Falls',
 'city_mpg': 19,
 'clickoff_url': '',
 'combined_mpg': 22,
 'condition': 'used',
 'cpl_value': 3,
 'created_at': '2023-06-30T02:01:50.807Z',
 'dealer_group_uuid': None,
 'dealer_name': 'Billion Chrysler Jeep Dodge Ram',
 'distance_from_origin': 91,
 'door_count': 4,
 'driveline': 'FWD',
 'eligible_for_financing': False,
 'email_opt_default': True,
 'engine_cylinders': 6,
 'entertainment': False,
 'experience': 'local',
 'finance_dealer_id': None,
 'financing_experience': 'none',
 'fuel_type': 'gasoline',
 'green_engine': False,
 'handicap_accessible': False,
 'has_photos': True,
 'heated_seats': True,
 'hwy_mpg': 27,
 'ipod_aux_input': False,
 'is_hot': False,
 'lane

In [31]:
data_jason['records'][0]['model_name']

'Edge'

In [32]:
data_jason['records'][0]['mileage']

101023

In [33]:
data_jason['records'][0]['city']

'Sioux Falls'

In [34]:
data_jason['records'][0]['dealer_name']

'Billion Chrysler Jeep Dodge Ram'

In [35]:
data_jason['records'][0]['year']

2013

In [36]:
data_jason['records'][0]['price']

13995

The required datas from the first page are scrapped and displayed

This website totally contains the details of used cars in `187` pages. Hence `multiple pages` are to be scrapped by using the same cURL

## Multiple page Web Scrapping

In [37]:
#multiple page scrapping for second city
headers = {
    'sec-ch-ua': '^\\^Not.A/Brand^\\^;v=^\\^8^\\^, ^\\^Chromium^\\^;v=^\\^114^\\^, ^\\^Google',
    'Referer': 'https://www.autolist.com/',
    'sec-ch-ua-mobile': '?0',
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36',
    'sec-ch-ua-platform': '^\\^Windows^\\^',
}

model = []
year = []
dealer = []
mileage = []
price = []
city=[]

for i in range(1,188):
    params = (
        ('include_total_price_change', 'true'),
        ('include_time_on_market', 'true'),
        ('include_relative_price_difference', 'true'),
        ('latitude', '42.8286091'),
        ('limit', '20'),
        ('longitude', '-98.3298021'),
        ('make', 'Ford'),
        ('page', '1'),
        ('radius', '100'),
        ('zip', '600099'),
    )

    response = requests.get('https://production-proxy.autolist.com/api/v2/search', headers=headers, params=params)

    #json object
    data_json=response.json()
    
    #getting 20 items for 187 page
    data_items=data_json['records']
    
    for data in data_items:
        model.append(data['model_name'])
        year.append(data['year'])
        dealer.append(data['dealer_name'])
        mileage.append(data['mileage'])
        price.append(data['price'])
        city.append(data['city'])

After scrapping off entire pages, the retrieved datas are stored in a `DataFrame` format named as `ford_pages2_df`.

## Converting into DataFrame

In [38]:
ford_pages2_df=pd.DataFrame({'Model': model,'Mileage': mileage,'City': city,'Year': year,'Dealer': dealer,'Price': price})

In [39]:
ford_pages2_df

Model   Mileage         City  Year  \
0                 Edge  101023.0  Sioux Falls  2013   
1                F-150    5600.0  Sioux Falls  2022   
2     F-250 Super Duty   91663.0       Neligh  2015   
3                F-150   93976.0       Neligh  2018   
4                F-150  248680.0     Niobrara  2016   
...                ...       ...          ...   ...   
3735             F-150       NaN     Mitchell  2023   
3736          Explorer   24150.0     Niobrara  2021   
3737             F-150      10.0     Mitchell  2023   
3738             F-150   73809.0  Sioux Falls  2018   
3739             F-150       NaN     Mitchell  2023   

                               Dealer  Price  
0     Billion Chrysler Jeep Dodge Ram  13995  
1     Billion Chrysler Jeep Dodge Ram  68995  
2           Jonny Dodge Chrysler Jeep  29450  
3           Jonny Dodge Chrysler Jeep  28450  
4                 Moody Motor Company  17995  
...                               ...    ...  
3735                   Vern Eide Ford  69705  
3736              Moody Motor Company  38995  
3737                   Vern Eide Ford  83565  
3738  Billion Chrysler Jeep Dodge Ram  43995  
3739                   Vern Eide Ford  78515  

[3740 rows x 6 columns]

The entire scrapped datas are a count of `3740` cars from `187` pages

Now the obtained dataframes for are both the cities are combined using `concat` function and stored as an `excel` sheet

## Merging of DataFrame 

In [40]:
#merging both dataframes
merged=pd.concat([ford_pages_df,ford_pages2_df],ignore_index=True)

In [41]:
merged

Model       Mileage         City  Year  \
0      Fusion Hybrid  52,618 Miles    Montclair  2018   
1      Fusion Energi  68,878 Miles    Montclair  2017   
2             Escape  20,032 Miles        Norco  2021   
3             Escape  60,292 Miles    Montclair  2018   
4             Fusion  54,599 Miles        Azusa  2018   
...              ...           ...          ...   ...   
13735          F-150           NaN     Mitchell  2023   
13736       Explorer       24150.0     Niobrara  2021   
13737          F-150          10.0     Mitchell  2023   
13738          F-150       73809.0  Sioux Falls  2018   
13739          F-150           NaN     Mitchell  2023   

                                                  Dealer    Price  
0                                Montclair Auto Exchange  $18,502  
1                                Montclair Auto Exchange  $17,413  
2      CarMax Norco - Now Offering Curbside Pickup & ...  $26,998  
3                                Montclair Auto Exchange  $16,324  
4                               California Motor Trade 2  $16,499  
...                                                  ...      ...  
13735                                     Vern Eide Ford    69705  
13736                                Moody Motor Company    38995  
13737                                     Vern Eide Ford    83565  
13738                    Billion Chrysler Jeep Dodge Ram    43995  
13739                                     Vern Eide Ford    78515  

[13740 rows x 6 columns]

In [42]:
merged.isnull().sum()

Model        0
Mileage    561
City         0
Year         0
Dealer       0
Price        0
dtype: int64

In [43]:
merged.dtypes

Model      object
Mileage    object
City       object
Year        int64
Dealer     object
Price      object
dtype: object

## Saving the scrapped data as Excel file

In [44]:
merged.to_excel('Ford.xlsx',index=False)

The entire details which were scrapped from `Autoline` website are stored in excel format named as `Ford`